Initial blog (#76)
grafana-operator/grafana-operator-experimental: docs/blog/v5-intro.md

# Version 5 RC

If you are reading this it means that we have finally got our first RC of version 5 out of the gates.

> The maintainers is extremely happy to finally being able to do so.

Reviewer: ```suggestion
The maintainers are extremely happy to finally being able to do so.
```

In this blog we will try to explain the changes that we have made and the future of the operator.
In [v5 getting started](v5-getting-started.md), you will be able to read how to try out the operator.

Version 5 or v5 as I will write from now on is a complete rewrite of version 4 of the operator.
The API versions are different so if you want you can run both the operators at the same time while you do a migration.

There are multiple reasons why we did this, but the main reason is the number of architectural changes that we have done.

Just a reminder this is an RC and not production ready, we will probably do a few breaking changes before releasing v5 but we really want to bring you something to try out and get some feedback on what we have done.

## Architectural changes

> The main change is that instead of grafana "finding" the dashboards, data sources CR they now find the grafana instance.

Reviewer: ```suggestion
The main change is that now, instead of grafana "finding" the dashboard and datasource CRs, they will now find the grafana instance.
```

Author: I got similar feedback from a friend, I rewrote it a bit. Can you take a second look and see if you think it looks better?

Reviewer: Yep, new version looks better, reading the original wording seemed a bit awkward, but I couldn't put my finger on what it was exactly, Thanks! :+1: :sweat_smile:

kind: Grafana
metadata:
  name: grafana
  labels:
    dashboards: "grafana"
spec:
  config:
    log:
      mode: "console"
    auth:
      disable_login_form: "false"
    security:
      admin_user: root
      admin_password: secret
```

And the grafana dashboard with the instanceSelector, the same thing goes for the other grafana resources.

```.yaml
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDashboard
metadata:
  name: grafanadashboard-sample
spec:
  resyncPeriod: 30s
  instanceSelector:
    matchLabels:
      dashboards: "grafana"
  json: >
    {
      "id": null,
      "title": "Simple Dashboard",
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
```

## New Features

> This is probably the part of the blog that you are most exited about, it goes through a number of the new features that we support.

Reviewer: ```suggestion
This is probably the part of the blog that you are most excited about, it goes through a number of the new features that we support.
```

### External grafana instances

It's now possible to configure external grafana instances through the operator.
So if you are using grafana cloud or a hosted grafana instances in one of the big cloud providers you can still use gitops to apply your dashboards to them.

One of the limitations with the external grafana instances is that we don't support installing plugins.
If your dashboard is using a plugin we assume that you have manually installed it to your grafana instance.

### Grafana config

The deployment configuration is more **opinionated** in it's default setup while at the same time giving you the user access to the whole deployment spec so you can overwrite anything you want.
For example when you start a grafana instance it will by default have decent security defaults, it will set `RunAsNonRoot` on your grafana instance when they get created.
But since you have access to the whole deployment spec you can always overwrite these kind of settings.

The same thing applies to all our resources, for example ingress.
The bad thing here is of course that we won't hide any of the logic that made the operator "easy" to get started with.

> So if you want to configure an ingress for your grafana instance you will have to define it just like any other ingress resources.

Reviewer 2: Refer to an example for this, like the normal nginx one and openshift route?

Author: I don't want to many example in this document, I added a ingress-nginx one. 
Instead I will focus on examples in the getting-started part.

But it makes it extremely extensible and hopefully this will make the operator support any setup you want.

### Grafana ini

Just like deployment config the same thing goes for grafana config, instead of having a CRD that contains all the grafana ini values we have instead made it possible to write any value as you want.
The **bad** thing is of course that you as an end user don't get any help with verifying types etc.
The **good** thing is that you can set any value you want, if you want to try out an experimental feature you can do so without first having to create a PR to the operator.

### Usage of status fields

The usability of v4 have more then once been laking. It's been really hard to understand why something is happening and you more or less have been forces to go in and look at the logs.
This is something that we wanted to change with v5.

In v5 you will be able to look at your grafana dashboard CR and see that it got a matching grafana instance (the label selector is correct) and if it's been applied to the grafana instances.

### No more plugin sidecar

> As you might know we have had a python container that have been an [sidecar](https://github.com/grafana-operator/grafana_plugins_init) to the operator that have been in charge of installing the plugins to grafana.

Reviewer 2: For future proofing, refer to a tag version of this repo if the plan is for v5 to be merged into this repo?

Author: So I'm a bit unsure what you meant, but in the top of the blog I have now added a mention on how we will manage the branches in the future. Version 5 will be the new mater and we will keep version 4 as a branch.

About grafana_plugins_init repo we will archive it when we stop supporting version 4 all together.
I hope this makes it clearer.

The sidecar have been badly maintained and to be honest it's just not needed. Instead we use an environment variable to install plugins during startup.

### Helm chart

We have decided to start to manage our own Helm chart. The main reason for this is that we want a simple way of giving you test versions of the operator and many of you use Helm to install the operator.

We still see a big place for the [community Helm chart](https://bitnami.com/stack/grafana-operator/helm) that bitnami is hosting and we hope that they will keep on doing so.

### Namespace mode

> Another issues that we have gotten lots of feedback on is [Excessive rbac access](https://github.com/grafana-operator/grafana-operator/issues/604)

Reviewer 2: /issues/issue/

This wasn't something that we could easily fix in v4 but in v5 we have really tried to accommodate this by introducing namespace mode.

> In namespace mode you can run the operator to only look for grafana resources in the same namespace that you are in or you can define it to look for a number of specific namespace. By default it will use cluster wide access.

Reviewer 2: second namespace should be plural?

If you install through Helm or Kustomize you will be able to define if you should create cluster roles or only roles.
Sadly for our Openshift users this won't be possible due to the limitations of OLM, even if you run the operator in namespace mode, which you can. OLM will still use the cluster wide rbac settings.

## v5

So there is still lots to do before version 5 will be released.

There are still a number of features that is missing and we have tried to create issues around this.

> Please have a look in our v5 [milestone](https://github.com/grafana-operator/grafana-operator/milestone/1) to see all open issues.

Author: This mileston don't exist, I hope I will have time to set this up later today.
I will start moving issues from the experimental repo to the grafana-operator one and from there I will label issues and add them to milestones.

But we need **your** help, first of all try out version 5.

- Do you like it?
- Is everything working as you expected?
- Is there any features missing?
- Is there an open issue for that feature? Please search for it and if it's not please create an issue.

We probably won't support all the features that we did in v4, due to maintenance burden, but if a missing feature get lots of up votes we might give it a second review.

An extra note about the **do you like it?**, getting feedback is one of the hardest things in open-source.
Even if "all" you did was to follow the new v5 [installation guide](v5-getting-started.md) we want to know about it, you might feel that you don't have specific insights or thoughts and that is fine.

> But writing one row on slack and saying "hey I installed version v5 and it worked as I expected". Helps us allot since we get to know that people are trying it.

Reviewer: ```suggestion
But writing one row on slack and saying "hey I installed version v5 and it worked as I expected". Helps us a lot since we get to know that people are trying it.
```

Author: This have been fixed.
